cpu/efm32/periph/adc_series2.c

```diff
@@ -0,0 +1,153 @@
+        configs.configs[i].osrHighSpeed = ADC_MODE_OSR(adc_config[dev].available_res[i]);
+        configs.configs[i].digAvg = ADC_MODE_AVG(adc_config[dev].available_res[i]);
+        configs.configs[i].analogGain = adc_config[dev].gain;
+        configs.configs[i].reference = adc_config[dev].reference;
+        configs.configs[i].vRef = adc_config[dev].reference_mV;
+        configs.configs[i].twosComplement = iadcCfgTwosCompUnipolar;
+    }
+    IADC_init(adc_config[dev].dev, &init, &configs);
+
+    return 0;
+}
+
+static inline GPIO_Port_TypeDef _port_num(gpio_t pin)
+{
+    return ((pin & 0xf0) >> 4);
+}
+
+static inline uint8_t _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+static void _setup_abus(gpio_t pin) {
+    const uint32_t alloc = GPIO_ABUSALLOC_AEVEN0_ADC0
+                         | GPIO_ABUSALLOC_AODD0_ADC0;
+    switch (_port_num(pin)) {
+        case gpioPortA:
+            GPIO->ABUSALLOC = alloc;
+            break;
+        case gpioPortB:
+            GPIO->BBUSALLOC = alloc;
+            break;
+        case gpioPortC:
+        case gpioPortD:
+            GPIO->CDBUSALLOC = alloc;
+            break;
+    }
+}
+
+int32_t adc_sample(adc_t line, adc_res_t res)
+{
+    uint8_t dev = adc_channel_config[line].dev;
+
+    /* find config to given resolution */
+    uint8_t config;
+    for (config = 0; config < IADC0_CONFIGNUM; config++) {
+        if (adc_config[dev].available_res[config] == res) {
+            /* we found the corresponding config */
+            break;
+        }
+    }
+    /* unsopported resolution */
+    if (config >= IADC0_CONFIGNUM) {
+        return -1;
+    }
+
+    /* lock device */
+    mutex_lock(&adc_lock[dev]);
+
+    /* setup channel and start sampling */
+    static const IADC_InitSingle_t init = {
+        .alignment = iadcAlignRight20,
+        .showId = false,
+        .dataValidLevel = iadcFifoCfgDvl4,
+        .fifoDmaWakeup = false,
+        .triggerSelect = iadcTriggerSelImmediate,
+        .triggerAction = iadcTriggerActionOnce,
+        .start = true
+    };
+    IADC_SingleInput_t input = IADC_SINGLEINPUT_DEFAULT;
+    input.configId = config;
+    input.posInput = IADC_portPinToPosInput(_port_num(adc_channel_config[line].input_pos),
+                                            _pin_num(adc_channel_config[line].input_pos));
+    _setup_abus(adc_channel_config[line].input_pos);
+    if (adc_channel_config[line].input_neg != GPIO_UNDEF) {
+        input.negInput = IADC_portPinToNegInput(_port_num(adc_channel_config[line].input_neg),
+                                                _pin_num(adc_channel_config[line].input_neg));
+        _setup_abus(adc_channel_config[line].input_neg);
+    }
+
+    IADC_initSingle(adc_config[dev].dev, &init, &input);
+
+    /* wait for the conservation to provide the result in the fifo */
+    while ((IADC_getStatus(adc_config[dev].dev) & IADC_STATUS_SINGLEFIFODV) == 0);
+    uint32_t result = IADC_pullSingleFifoData(adc_config[dev].dev);
```

Comment:
If you set a negative input for differential mode, can't the measured voltage be negative?

You'd want a `int32_t` result then.

Reply:
https://github.com/RIOT-OS/RIOT/blob/d759701030a076363a20ca6713735f1f80c1640e/cpu/efm32/periph/adc_series2.c#L68

This line makes sure that the result is never negative.

I took this route to be consistent with the `adc_sample()` interface which uses negative values to indicate errors.

Reply:
> I took this route to be consistent with the `adc_sample()` interface which uses negative values to indicate errors.

on sam0 I just ignored that interface convention - if there is a wrong config the function should just throw an `assert()` failure.

`adc_sample()` just doesn't do this because the test does use this to query supported resolutions, but I think no real-world application will do that.

There should just be a separate interface to check if a resolution is valid (adc_ng does this).

Reply:
Hmm, I don't want to mess around with the defined (and of course very limited) ADC interface.

Otherwise the user can't tell whether [something wrong](https://github.com/RIOT-OS/RIOT/blob/64dcfd67ff3b8368bae714e098c2b2b316f43883/cpu/sam0_common/periph/adc.c#L293-L296) or the ADC just [measured `-1`](https://github.com/RIOT-OS/RIOT/blob/64dcfd67ff3b8368bae714e098c2b2b316f43883/cpu/sam0_common/periph/adc.c#L330). Instead, I'd would wait for an ADC interface that's capable of reporting negative values ...

Reply:
What does `iadcCfgTwosCompUnipolar` mean?
Will the application have to deal with this then?

Reply:
It clamps the value to `0` if it would become negative. cf. https://docs.silabs.com/gecko-platform/3.0/emlib/api/efr32xg22/group-iadc#gadaad7387cbc0b5a22c631d54fa01d70c

Reply:
We should add a `adc_check_res(adc_t line, adc_res_t res)` API so that `adc_sample()` will never need to return error.

Reply:
Huh? There's already an API for that?

Reply:
I only see `adc_init(adc_t line)` and `adc_sample(adc_t line, adc_res_t res)` in `periph/adc.h`

Reply:
Ah, it's an idea for a follow-up? We should have a deeper look into `adc_ng` - maybe that's a good abstraction to fit this case. Otherwise we have the go through the API change process ...